homeassistant/components/compensation/__init__.py

```diff
@@ -0,0 +1,132 @@
+"""The Compensation integration."""
+import logging
+import re
+import warnings
+
+import numpy as np
+import voluptuous as vol
+
+from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN
+from homeassistant.const import (
+    CONF_ATTRIBUTE,
+    CONF_ENTITY_ID,
+    CONF_NAME,
+    CONF_UNIT_OF_MEASUREMENT,
+)
+from homeassistant.helpers import config_validation as cv
+from homeassistant.helpers.discovery import async_load_platform
+
+from .const import (
+    CONF_COMPENSATION,
+    CONF_DATAPOINTS,
+    CONF_DEGREE,
+    CONF_POLYNOMIAL,
+    CONF_PRECISION,
+    DATA_COMPENSATION,
+    DEFAULT_DEGREE,
+    DEFAULT_NAME,
+    DEFAULT_PRECISION,
+    DOMAIN,
+    MATCH_DATAPOINT,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def datapoints_greater_than_degree(value: dict) -> dict:
+    """Validate data point list is greater than polynomial degrees."""
+    if not len(value[CONF_DATAPOINTS]) > value[CONF_DEGREE]:
+        raise vol.Invalid(
+            f"{CONF_DATAPOINTS} must have at least {value[CONF_DEGREE]+1} {CONF_DATAPOINTS}"
+        )
+
+    return value
+
+
+COMPENSATION_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_ENTITY_ID): cv.entity_id,
+        vol.Required(CONF_DATAPOINTS): vol.All(
+            cv.ensure_list(cv.matches_regex(MATCH_DATAPOINT)),
```

Comment:
This seems a bit silly. Just have it be a list of pairs?

```yaml
datapoints:
  - [0, 55]
  - [100, 200]
```

Reply:
This is completely untested, but something like:
```
vol.Required(CONF_DATAPOINTS): [
 vol.ExactSequence([vol.Coerce(float), vol.Coerce(float]])
]
```

Reply:
I was matching the compensation applied with ESPHome's configuration because it would be familiar.  I also didn't want them to think of it as an XY point.

Reply:
![image](https://user-images.githubusercontent.com/35082313/108910263-9b6c4980-75f3-11eb-805a-edc80ded934a.png)

Reply:
The more I think about this, the more I disagree with changing it.  If we go with your method, the yaml will look really odd:
```
datapoints:
  - 
    - 0
    - 55
  - 
    - 100
    - 200
```

Reply:
If we ever want to move forward by adding UI capabilities to this integration, I would really recommend sticking to regular data structures.

From that perspective, I agree with @elupus:

```yaml
datapoints:
  - [0, 55]
  - [100, 200]
```

Reply:
Not sure that need to be argument. The voluptuous schema could mutate the string into standard datapoint on validation. Perhaps best of both worlds?

Reply:
FWIW I had config flow in this originally but removed it because it didn't work the way I wanted.  I was able to use the current validation inside it and it worked great.  Was even able to highlight which points that were formatted incorrectly.  The reason I removed config flow was because of the following:

1.  You can't dynamically populate a combobox based on the result of another combobox inside the same config flow window.
2.  I couldn't add a button or action that would append or remove a row inside the same config flow window.

So keeping it the way it is shouldn't be a problem when moving to the UI if/when those 2 abilities become available.

Either way, I can make the change to a pair of doubles, I just think it will be more confusing to users who don't know what they are doing.

Reply:
> Not sure that need to be argument. The voluptuous schema could mutate the string into standard datapoint on validation. Perhaps best of both worlds?

It is, as it allows for possible number inputs instead of free form strings.

Reply:
What i meant was.

vol.Any(regexp_match_mutate, standard_data) -> always standard lists stored in the resulting config structures.

Reply:
I get that, I was not talking about anything backend. My original comment:

> If we ever want to move forward by adding UI capabilities to this integration

It was aimed at the UI experience. Showing correct input that matches the data is a good practice generally, or else, everything could as well be a text area on the internet.

That we can coerce on the backend, I understand and was not a concern for me.